File: Andriod/Practice/app/src/main/java/com/example/practice/Adapter/ViewPagerAdapter.java
package com.example.practice.Adapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;

import androidx.annotation.NonNull;
import androidx.viewpager.widget.PagerAdapter;

import com.example.practice.R;

import java.util.ArrayList;

public class ViewPagerAdapter extends PagerAdapter {

    private Context context;
    private ArrayList<Integer> listImage;
    public ViewPagerAdapter(Context context, ArrayList<Integer> listImage) {
        this.context=context;
        this.listImage=listImage;
    }


    @NonNull
    @Override
    public Object instantiateItem(@NonNull ViewGroup container, int position) {
        LayoutInflater inflater=LayoutInflater.from(context);
        View view=inflater.inflate(R.layout.row_viewpager,container,false);
        ImageView imageView;
        LinearLayout layout;
        imageView=view.findViewById(R.id.imageAdapter);
        imageView.setImageResource(listImage.get(position));
        container.addView(view, 0);
        return view;
    }


    @Override
    public int getCount() {
        return listImage.size();
    }

    @Override
    public boolean isViewFromObject(@NonNull View view, @NonNull Object object) {
        return view==object;
    }


    @Override
    public void destroyItem(ViewGroup collection, int position, Object view) {
        collection.removeView((View) view);
    }
}

File: Andriod/PracticeProject/app/src/main/java/com/ali/sampleandroid/adapter/ViewpagerAdapter.kt
package com.ali.sampleandroid.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.viewpager.widget.PagerAdapter
import com.ali.sampleandroid.databinding.RowPagerBinding

class ViewpagerAdapter(private val listImages: List<Int>,private val listener: (Int) -> Unit) : PagerAdapter() {

    //It returns the no of view available
    override fun getCount(): Int {
        return listImages.size
    }

    override fun isViewFromObject(view: View, obj: Any): Boolean {
        return view == obj
    }

    override fun instantiateItem(container: ViewGroup, position: Int): Any {
        /*
        * Before view binding
        val imageLayout = LayoutInflater.from(container.context).inflate(R.layout.row_pager,container,false)
        val imageView = imageLayout.findViewById<ImageView>(R.id.ivPager)
        imageView.setImageResource(listImages[position])
        container.addView(imageLayout,0)
        return imageLayout*/

        //Using view binding
        val binding = RowPagerBinding.inflate(LayoutInflater.from(container.context),container,false)
        binding.ivPager.setImageResource(listImages[position])
        binding.ivPager.setOnClickListener {
            listener(position)
        }
        container.addView(binding.root,0)
        return binding.root
    }

    override fun destroyItem(container: ViewGroup, position: Int, obj: Any) {
        container.removeView(obj as View)
    }
}